### PR TITLE
[Feature] toFactory implementation

### PR DESCRIPTION
**Pull request by SieR-VR**
- [x] implement #69

### CODE REVIEW

File: src/async.ts
Comment (willryan) [inline src/async.ts:18]: I guess this change was made to keep `lift()` happy? But `isPromise()` will now compile with null or undefined as arguments, and at present that will cause this function to crash.  maybe change the function body to:

```typescript
  return !!t && typeof (t as any)["then"] === "function";
```

File: src/async.ts
Comment (willryan) [inline src/async.ts:276]: I generally don't like reassignment like this, although Typescript does helpfully update the known type of `builder`.  The `lift()` function is meant to help with situations like this:
```
const resolvedBuilder = await lift(builder);
// now use resolvedBuilder going forward
```

File: src/pipeline.ts
Comment (willryan) [inline src/pipeline.ts:96]: I don't love that we have to cast here, but I can't find anything more elegant.  `this.current` _is_ a valid `Async.Builder<P>`, but unfortunately makeFactory's signature won't quite allow it.  Maybe make a comment here explaining why the cast is valid?  Also you can just say `Promise<Async.Builder<P>>` (the second generic ar defaults to `keyof P`)

**Review by willryan — CHANGES_REQUESTED** [review]
Really great work!  Thanks for taking this on, and sorry it took me a bit to get to it.  I did finally get approval to keep working on this project under my full ownership, though, so that's good.  I just a few small requests / suggestions.

File: src/pipeline.ts
Comment (SieR-VR) [reply src/pipeline.ts]: this.current is Promise object, so I cannot help to use cast as Promise<Async.Builder<P, keyof P>>.
and I changed type parameter, thanks for review.

**Review by willryan — APPROVED** [review]
Looks good.  I'll push out a new version after this gets merged.

**Pull request merged**
